IO forward-compatibility feature flags

bbockelm opened the pull request:
This PR introduces (with unit tests!) the concept of a forward-compatibility feature break flag for `TBasket`.

This allows the `TBasket` class to identify if the object being deserialized was written by a newer version of ROOT using a feature that breaks forward-compatibility.  Note that much care was taken so older versions of ROOT that *don't* recognize the new flag will believe the file is corrupt; they will give a misleading error message, but will not silently serve corrupted data.

If this approach goes forward, I intend to backport the flags to older versions of ROOT -- they should be able to recognize files they aren't supposed to read.  I do not plan to backport the new IO features themselves.

pcanal commented:
@bbockelm Your code seems to trigger a deficiency in TStatusBitsChecker.  Can you take a  look of why it crashes?  thanks.

pcanal commented on tree/tree/inc/TBasket.h:
```diff
@@ -35,6 +34,8 @@ class TBranch;
 
 class TBasket : public TKey {
 
+   friend class TBasket_TestUnsupportedIO_Test;
```
Is this used?

bbockelm replied:
Yes - unit test below.  `gtest` style is to include `gtest/gtest_prod.h` and do:

```
FRIEND_CLASS(TBasket, TestUnsupportedIO)
```

However, I didn't want to intermix `gtest` headers and ROOT public APIs...

pcanal replied:
> However, I didn't want to intermix gtest headers and ROOT public APIs...

Yes, that would be worse :) ...

pcanal commented on tree/tree/src/TBranch.cxx:
```diff
@@ -1163,6 +1163,9 @@ TBasket* TBranch::GetBasket(Int_t basketnumber)
       }
       Error("GetBasket","File: %s at byte:%lld, branch:%s, entry:%lld, badread=%d, nerrors=%d, basketnumber=%d",file->GetName(),basket->GetSeekKey(),GetName(),fReadEntry,badread,nerrors.load(),basketnumber);
+      if (basket->IsZombie()) {
+         MakeZombie();
```
This is too strong of statement.  In the usual case, there is some corruption on the file and **some** but not all the basket can be read.

pcanal commented on tree/tree/test/TBasket.cxx:
```diff
@@ -0,0 +1,146 @@
+   ASSERT_FALSE(tree->IsZombie());
+
+   TBranch *br = tree->GetBranch("idx");
+   ASSERT_TRUE(br != nullptr);
```
why not:
```
ASSERT_NE(br, nullptr);
```
?

pcanal commented on tree/tree/test/TBasket.cxx:
```diff
@@ -0,0 +1,146 @@
+   // When we are down to one bitset, we'll have to expand the field.
+   EXPECT_TRUE(static_cast<UChar_t>(~TBasket::kSupported));
+
+   basket->fIOBits = ~TBasket::kSupported;
```
Shouldn't we have TBasket::GetIOBits() instead? (and/or TestIOBits(TBasket::EIOBits))

pcanal replied:
Humm ... we do ... but I see ... this is an assignment here ... You could also do it by grabbing the offset of fIOBits via TClass/TDataMember.

bbockelm replied:
I want to avoid providing interfaces to use the IOBits infrastructure in this PR (aim is to keep backports simple -- no new API is introduced into older versions of ROOT if we backport this).

While the `TClass`/`TDataMember` technically could be done -- is it really more straightforward than just allowing the unit test to be a friend class?

pcanal replied:
It is relatively straightforward but indeed not as straightforward as the friend class.

However I dislike the spurrious friend declaration .. and 'more' importantly the name of the class is actually dependent on the gtest implementation.  They may or may not change their naming algorithm in a future version.

amadio commented:
@pcanal The status bit check test seems a bit flaky as it also failed in the [nightlies](http://cdash.cern.ch/viewTest.php?onlyfailed&buildid=394075), albeit with a different problem (segfault, not error message).

bbockelm commented on tree/tree/src/TBasket.cxx:
```diff
@@ -769,24 +771,21 @@ void TBasket::SetWriteMode()
       b >> fBufferSize;
       b >> fNevBufSize;
       if (fNevBufSize < 0) {
-         Error("Streamer","The value of fNevBufSize is incorrect (%d) ; trying to recover by setting it to zero",fNevBufSize);
```
Yes, it is replaced by https://github.com/root-project/root/pull/959/files#diff-244ab93230dcfa0c91eb218428db7be6R815

Basically, if you have a negative value here (corruption of the upper bit), then we apply  checks to the serialized `fIOBits` -- the upper bit there also must be zero.

So, before, we can detect a single bit flip; afterward, we can detect a single bit flip.

pcanal replied:
>  the upper bit there also must be zero.

Humm ... this is not documented, is it?  I.e. we 'might' have 255 variations (we won't but nothing yet prevent the use of the upper bit :)

> So, before, we can detect a single bit flip; afterward, we can detect a single bit flip.

Fair enough, let's make it clear that the location of the test switched in comments or commits.

pcanal replied:
One of the primary side effect was to avoid executing:
```
tree/tree/src/TBasket.cxx:   if (fNevBufSize) {
tree/tree/src/TBasket.cxx:      fEntryOffset = new Int_t[fNevBufSize];
```
with a crazy fNevBufSize value.  I did not follow the new code enough (with your new test on IsZombie) to verify whether the execution of that line is avoided in case of seeming corruption.

> So, before, we can detect a single bit flip; afterward, we can detect a single bit flip.

Actually not quite.  With a file that **is** using the extensions, a crazy value of fNevBufSize will no longer be detected and can still lead to a large memory allocation (half as large as before but still ....)

bbockelm commented on tree/tree/src/TBasket.cxx:
```diff
@@ -811,6 +810,23 @@ void TBasket::Streamer(TBuffer &b)
+            MakeZombie();
+            fNevBufSize = 0;
+         } else if (fIOBits && (fIOBits & ~kSupported)) {
+            Error("TKey::Streamer", "The value of fIOBits (%s) contains unknown flags (supported flags "
```
I make the effort to `MakeZombie`-fy the TBranch.  Is that sufficient?

pcanal replied:
> I make the effort to MakeZombie-fy the TBranch. Is that sufficient?

Well no.  TBranch itself does not check it and thus will still try to read.  [Also as it is now it is set too often]

bbockelm commented on tree/tree/src/TBasket.cxx:
```diff
@@ -811,6 +810,23 @@ void TBasket::Streamer(TBuffer &b)
+         } else if (fIOBits && (fIOBits & ~kSupported)) {
+            Error("TKey::Streamer", "The value of fIOBits (%s) contains unknown flags (supported flags "
+                                    "are %s), indicating this was written with a newer version of ROOT "
+                                    "utilizing critical IO features this version of ROOT does not support."
```
Ah - I mean `critical` in the technical stance (as in X509 "critical extensions"): an error should be thrown if a feature is used and this version of ROOT doesn't understand it.

bbockelm commented on tree/tree/inc/TBasket.h:
```diff
@@ -61,6 +63,21 @@ class TBasket : public TKey {
+      // kSupported = kGenerateOffsetMap | kBasketClassMap
+      // kGenerateOffsetMap = BIT(1),
+      // kBasketClassMap = BIT(2),
+      kSupported = 0
```
I'm not sure it's a foregone conclusion that `kSupported` should be `|` of all EIOBits.  Particularly, if there is an experimental feature that we decide to not put into production, we want the ability to define a IOBit but not support it.

Now, in a future PR when I add an interface in the `ROOT::Experimental` namespace to enable / set IOBits in a basket, we must include a check that the user-requested IOBit is supported.

pcanal replied:
> Particularly, if there is an experimental feature that we decide to not put into production, we want the ability to define a IOBit but not support it.

This can be handled by adding a EObsoleteIOBits.  This has the additional great advantage of documenting for the user too the status of such bits (defined but not supported).

bbockelm replied:
Done!

pcanal commented:
> The status bit check test seems a bit flaky as it also failed in the nightlies, albeit with a different problem (segfault, not error message).

@amadio humm ... Odd ... I can't find any recent nightly showing the problem.  Can you point me in the right direction?   Also odd, *this* failure is a segfault and not an error message (aka I am confused by your message ;) ).

pcanal commented on tree/tree/src/TBasket.cxx:
```diff
@@ -769,24 +771,21 @@ void TBasket::SetWriteMode()
+   bool hasIOBits = false;
    if (b.IsReading()) {
       TKey::Streamer(b); //this must be first
       Version_t v = b.ReadVersion();
```
We should also increase the version of the TBasket class.

bbockelm commented:
@pcanal - I believe I have addressed all review comments.

I cannot reproduce the failure of `roottest-root-core-execStatusBitsCheck`; it's possible the fact I switched to an `enum class` avoids the underlying bug in that test?

pcanal commented on tree/tree/inc/TBasket.h:
```diff
@@ -52,6 +51,8 @@ class TBasket : public TKey {
    Int_t       fLast;            ///< Pointer to last used byte in basket
    Bool_t      fHeaderOnly;      ///< True when only the basket header must be read/written
+   UChar_t fIOBits{
+      0}; ///<!IO feature flags.  Serialized in custom portion of streamer to avoid forward compat issues unless needed.
```
Please align the variable names :)

bbockelm replied:
Should I listen to you or to `clang-format` here?

pcanal replied:
Nope :)  It can't do what we need here ...

pcanal commented on tree/tree/inc/TBasket.h:
```diff
@@ -60,6 +61,28 @@ class TBasket : public TKey {
+   // If (fIOBits & ~kSupported) is non-zero -- i.e., an unknown IO flag is set
+   // in the fIOBits -- then the zombie flag will be set for this object.
+   //
+   enum class EIOBits {
```
Please specify the storage to be used.
```
enum class EIOBits: char
```

pcanal commented on tree/tree/inc/TBasket.h:
```diff
@@ -60,6 +61,28 @@ class TBasket : public TKey {
+   // (kUnsupported | kSupported) should result in the '|' of all IOBits.
+   enum class EUnsupportedIOBits { kUnsupported = 0 };
+   // The number of known, defined IOBits.
+   static const int kIOBitCount = 0;
```
I recommend a constexpr ....

pcanal commented:
> I cannot reproduce the failure of roottest-root-core-execStatusBitsCheck; it's possible the fact I switched to an enum class avoids the underlying bug in that test?

It is indeed possible.  It could also be a platform dependent thing or it could be depend on the library actually build as part of ROOT ....

pcanal commented:
Last step :) ... we need to add a nice description of this feature in at least the release notes.

bbockelm commented:
@pcanal - release notes updated.

I tried to reproduce the above test failure locally, but was unable to.  Do you recognize it?

vgvassilev commented on tree/tree/test/TBasket.cxx:
```diff
@@ -0,0 +1,198 @@
+
+#include "gtest/gtest.h"
```
Could you sort headers by genrality, i.e. By moving gtest after the ROOT headers?

bbockelm commented:
@pcanal - are there any remaining items for this PR?

pcanal commented on README/ReleaseNotes/v612/index.md:
```diff
@@ -60,6 +61,8 @@ auto h1 = key->ReadObject<TH1>
 or will be set to the address of the histogram read from the file.
 
+- Added a new mechanism for providing clean forward-compatibility breaks in a ``TTree`` (i.e., a newer version of ROOT writes a ``TTree`` an older version cannot read).  When future versions of ROOT utilize an IO feature that this version does not support, ROOT will provide a clear error message instead of crashing or returning garbage data.  In future ROOT6 releases, forward-compatibility breaks ...
+
```
Please add example of failure both for unpatched and patch version.  thanks.

bbockelm replied:
This has been addressed by c7f8939.

bbockelm commented:
Ah - I see this also got tripped up by the movement of `TRangeStaticCast`.  Working on a rebase...

pcanal commented:
Please no merge commits :)

bbockelm commented:
@pcanal - rebase has been pushed!
